### PR TITLE
Fix API and ABI compatibility for old Androids

### PR DESCRIPTION
**Pull request by bassosimone**
xref: #376 
cc: @lorenzoPrimi

### CODE REVIEW

File: build/android/dependency [hunk 1]
```diff
@@ -12,23 +12,18 @@ ANDROIDDIR=$BUILDDIR/android
 
 NDK_DIR=$1
 SPEC=$2
-API=21
-if [ "$3" != "" ]; then
-    API=$3
-fi
```
Comment (bassosimone) [inline build/android/dependency]: Now we use the lowest possible API available for the architecture.

File: build/android/dependency [hunk 2]
```diff
@@ -12,23 +12,18 @@ ANDROIDDIR=$BUILDDIR/android
...
 
 echo ""
 echo "Cross-compiling $SPEC for all archs"
 
-ALL_ARCHS="aarch64-linux-android arm-linux-androideabi arm-linux-androideabi-v7a mipsel-linux-android mips64el-linux-android x86 x86_64"
+ALL_ARCHS="armeabi armeabi-v7a arm64 mips mips64 x86 x86_64"
```
Comment (bassosimone) [inline build/android/dependency]: The script to generate the toolchain has changed (the bash one has been deprecated for quite some time in favor of the python one that we're using now); architecture names changed accordingly

File: build/android/dependency [hunk 3]
```diff
@@ -12,23 +12,18 @@ ANDROIDDIR=$BUILDDIR/android
...
 
 for ARCH in $ALL_ARCHS; do
     echo ""
     echo "Cross compiling $SPEC for arch ${ARCH}"
-    if [ ! -d $ANDROIDDIR/toolchain/${ARCH}-${API} ]; then
-        $ANDROIDDIR/make-toolchain ${NDK_DIR} ${ARCH} ${API}
-    fi
-    $ANDROIDDIR/cross ${ARCH} ${API} $BUILDDIR/dependency $SPEC
+    rm -rf $ANDROIDDIR/toolchain/${ARCH}
+    $ANDROIDDIR/make-toolchain ${NDK_DIR} ${ARCH}
```
Comment (bassosimone) [inline build/android/dependency]: Do not cache the toolchain. Zap and recreate every time.

File: build/android/dependency [hunk 4]
```diff
@@ -12,23 +12,18 @@ ANDROIDDIR=$BUILDDIR/android
...
+    $ANDROIDDIR/cross ${ARCH} $BUILDDIR/dependency $SPEC
```
Comment (bassosimone) [inline build/android/dependency]: Not passing around API anymore.

File: build/android/make-toolchain [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
+    echo "  ARCH: armeabi, armeabi-v7a, arm64, mips, mips64, x86, x86_64" 1>&2
+    echo "  API: 1 ... 25" 1>&2
+    echo "" 1>&2
+    echo "If API is omitted, the minimum possible API for the " 1>&2
+    echo "required architecture is used." 1>&2
+    echo "" 1>&2
+    echo "Example usage (on MacOS with Android Studio installed):" 1>&2
+    echo "  $0 ~/Library/Android/sdk/ndk-bundle arm 19" 1>&2
+    echo "" 1>&2
     exit 1
```
Comment (bassosimone) [inline build/android/make-toolchain]: Now API is optional. If not specified we use the lowest possible API for architecture.

File: build/android/make-toolchain [hunk 6]
```diff
@@ -18,27 +25,27 @@ ARCH=$2
...
+if [ "$API" = "" ]; then
+    if [ "$ARCH" = "arm64" -o "$ARCH" = "mips64" -o "$ARCH" = "x86_64" ]; then
+        API=21
+    else
+        API=9
+    fi
+fi
 
-MAKE_TOOLCHAIN=${NDK_DIR}/build/tools/make-standalone-toolchain.sh
+MAKE_TOOLCHAIN=${NDK_DIR}/build/tools/make_standalone_toolchain.py
```
Comment (bassosimone) [inline build/android/make-toolchain]: The new script (the old one is deprecated)

File: build/android/cross [hunk 7]
```diff
@@ -3,36 +3,24 @@ set -e
...
     echo "Example usage:" 1>&2
-    echo " $0 aarch64-linux-android 21" 1>&2
+    echo " $0 arm64" 1>&2
     exit 1
 fi
 
 ARCH=$1
 shift 
-API=$1
-shift
```
Comment (bassosimone) [inline build/android/cross]: Now we use lowest API for architecture

File: build/android/cross [hunk 8]
```diff
@@ -3,36 +3,24 @@ set -e
...
-# XXX shortcut for armeabi-v7a
-# According https://developer.android.com/ndk/guides/standalone_toolchain.html
-# the proper solution is to override CFLAGS
-if [ $ARCH = "arm-linux-androideabi-v7a" ]; then
-    echo "WARNING: converting $ARCH to arm-linux-androideabi" 1>&2
-    ARCH="arm-linux-androideabi"
-    V7A="-v7a"
-else
-    V7A=""
-fi
```
Comment (bassosimone) [inline build/android/cross]: This diff also improves the way in which we compile for arm. We now pass the same flags as ndk-build and we deal properly with both armv5 and armv7.

**Comment by lorenzoPrimi** [comment]
Great, finally something working on 99% of android devices! You can merge

**Pull request merged**